designs/dynamic-bundle.md

```diff
@@ -0,0 +1,267 @@
+Dynamic Bundle
+===================
+
+## Overview
+Kubebuilder currently has an implementation of a `Bundle` which allows multiple plugins to be *bundled* together. The execution of the plugins within a `Bundle` follows a sequential order based on their definition. 
+
+An example of the current method:
+
+```go
+gov3Bundle, _ := plugin.NewBundle(golang.DefaultNameQualifier, plugin.Version{Number: 3},
+	kustomizecommonv1.Plugin{},
+	golangv3.Plugin{},
+)
+```
+
+In the above example a plugin `Bundle` is created that ensures that the `kustomize/v1` plugin is run before the `go/v3` plugin.
+ 
+With the addition of Phase 2 Plugins on the way, users using the `--plugins` flag to utilize external plugins will not get the benefits of using some default functionality that previous Kubebuilder users using the default `go/v3` plugin gained via the `kustomize/v1` plugin.
+
+## Proposal
+The proposed solution is to create a new form of `Bundle` that is more dynamic and can be used to create a default bundle that contains sensible default plugins that can be used with all plugins. This could be overriden by a command line flag.
+
+## User Stories
+- As a Kubebuilder maintainer I would like to have Kubebuilder implement sane default plugins that always run before and/or after plugins, specifying this once and it applying to all uses
+
+- As a Kubebuilder user I expect Kubebuilder to have default functionality to make my operator development easier
+
+- As a Kubebuilder user I would like to be able to override the default Kubebuilder plugin chain, scaffolding only exactly what I feel I need
+
+## Goals
+- `kubebuilder` is able to define sane defaults for plugins that should always be run before and/or after user specified plugins
+
+- A `Bundle` no longer has to be created for each plugin to implement sane defaults
+
+## Examples
+Context: Kubebuilder defines that the `kustomize/v1` plugin should always run before any other plugin. Kubebuilder also defines the default plugin to be run if a user does not specify plugins via the `--plugins` flag is `go/v3`
```

Comment:
We have the config-gen alpha command which is a solution that replaces kustomize
When it was implemented to the project the idea was users could choose the best scaffold for their needs
So that, they could have a new Golang plugin that uses kustomize or config-gen.

If we move with this idea would that mean remove config-gen and accepted that the only way to configure the project is with kustomize. Am I right? is that something that we really want? Why we would say that a specific plugin is required for any based action done with the CLI? What are pros and cons here? How that can affect others? What is the plan to deprecate and remove config-gen (if that is the case)?

Reply:
Thank you. @everettraven it seems that all pros are already addressed with the current implementation. So, I still trying to figure out what we can do with the proposal that we cannot do today?

Reply:
> The API allows define the default plugins for each project version indeed, see: https://github.com/operator-framework/operator-sdk/blob/master/internal/cmd/operator-sdk/cli/cli.go#L122-L124

My understanding is that those plugins are ones to be run if a user does not use the `--plugins` flag. 

The proposal here is to implement a bundle that no matter what plugins a user passes in with the `--plugins` flag certain plugins are still run unless the user specifically wants to override that functionality. I think of it more as a way to have only one bundle definition that will apply for multiple plugins rather than a new bundle definition for each plugin to implement the same defaults.